### PR TITLE
[#1178] feat(trino-connector) Support create/drop catalog in Trino.

### PR DESCRIPTION
**Pull request by diqiu50**
### What changes were proposed in this pull request?

Support create/drop catalog in Trino.

```text
-- Call stored procedures with position.
call gravitino.system.create_catalog(
    'mysql',
    'jdbc-mysql',
    Map(
        Array('jdbc-url', 'jdbc-user', 'jdbc-password', 'jdbc-driver'),
        Array('jdbc:mysql:192.168.164.4:3306?useSSL=false', 'trino', 'ds123', 'com.mysql.cj.jdbc.Driver')
    )
)
call gravitino.system.drop_datalog('mysql');

-- Call stored procedures with name.
call gravitino.system.create_catalog(
    catalog =>'mysql',
    provider => 'jdbc-mysql',
    properties => Map(
        Array('jdbc-url', 'jdbc-user', 'jdbc-password', 'jdbc-driver'),
        Array('jdbc:mysql:192.168.164.4:3306?useSSL=false', 'trino', 'ds123', 'com.mysql.cj.jdbc.Driver')
    ),
    ignore_exist => true
)

call gravitino.system.drop_datalog(
    catalog => 'mysql'
    ignore_not_exist => true
);
```

### Why are the changes needed?

Feature: #1178

### How was this patch tested?

UT and IT

### CODE REVIEW

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java [hunk 1]
```diff
@@ -44,6 +47,12 @@ public String getMetalake() {
     return config.getOrDefault(GRAVITINO_METALAKE.key, GRAVITINO_METALAKE.defaultValue);
   }
 
+  boolean isStatic() {
+    // The 'isStatic' is indicated whether the connector
+    // is user-configured within Trino or loaded from the Graviton server
+    return config.getOrDefault("gravitino.internal", "false").equals("false");
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java:53]: Please also change the configuration document for the Trino connector.

[the remark this answers is not in the document]
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java:53]: No, you don't need to set gravitino.internal. That's not a configuration.
It's only generated in `CatalogInjector.createCatalogProperties`.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java [hunk 2]
```diff
@@ -66,17 +70,26 @@ public Connector create(
...
+    if (config.isStatic()) {
+      // Default GravitinoConnector named "gravitino" is just using to load
+      // CatalogConnectorManager,
+      // that's dynamically loading catalogs from gravitino server.
+
+      String metalake = config.getMetalake();
+      if (Strings.isNullOrEmpty(metalake)) {
+        throw new TrinoException(GRAVITINO_METALAKE_NOT_EXISTS, "No gravitino metalake selected");
+      }
+      catalogConnectorManager.addMetalake(metalake);
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java:85]: We can create multiple configuration files in the Trino catalog directory. Subsequently, we can have two Gravitino connectors to use two metalakes.

For example:

```
root@trino-01:/usr/local/trino-server-426/etc/catalog# ls
gravitino.properties  gravitino1.properties  jmx.properties  mysql.properties
```

gravitino.properties :
```
connector.name=gravitino
gravitino.url=http://localhost:8090
gravitino.metalake=test
```
gravitino1.properties 
```
connector.name=gravitino
gravitino.url=http://localhost:8090
gravitino.metalake=test1
```

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java [hunk 3]
```diff
@@ -103,20 +109,21 @@ public void start() {
...
-      try {
-        metalake = gravitinoClient.loadMetalake(NameIdentifier.ofMetalake(usedMetalake));
-      } catch (NoSuchMetalakeException noSuchMetalakeException) {
-        LOG.warn("Metalake {} does not exist.", usedMetalake);
-        return;
-      } catch (Exception e) {
-        LOG.error("Load Metalake {} failed.", usedMetalake, e);
-        return;
-      }
+      for (String usedMetalake : usedMetalakes) {
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:112]: The typical metadata lake usually has only one or two. For now, we don't need to consider it.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java [hunk 4]
```diff
@@ -103,20 +109,21 @@ public void start() {
...
+          continue;
+        } catch (Exception e) {
+          LOG.error("Load Metalake {} failed.", usedMetalake, e);
+          continue;
+        }
 
-      LOG.debug("Load metalake: " + usedMetalake);
-      loadCatalogs(metalake);
-      // TODO (yuhui) need to handle metalake dropped.
+        LOG.debug("Load metalake: " + usedMetalake);
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:124]: It's not advisable to use `debug` level here.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java [hunk 5]
```diff
@@ -44,6 +47,12 @@ public String getMetalake() {
     return config.getOrDefault(GRAVITINO_METALAKE.key, GRAVITINO_METALAKE.defaultValue);
   }
 
+  boolean isStatic() {
+    // The 'isStatic' is indicated whether the connector
+    // is user-configured within Trino or loaded from the Graviton server
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java:52]: Graviton -> Gravitino

**Comment by diqiu50** [comment]
@jerryshao  Please take a review.

**Comment by yuqi1129** [comment]
> hao Please take a review.

I 've already reviewed it. It's fine with me. Please take some time to view it. Thanks.

**Comment by diqiu50** [comment]
> With this, does user still need to configure metalake in trino connector?

Yes，the Trino connector will only load the configured metadata lakes."

**Comment by jerryshao** [comment]
So it the configured metalake is "test", and with call procedure you created a new one "test1", can "test1" be used by this connector, or user can still use the metalake "test"?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java [hunk 6]
```diff
@@ -20,8 +30,85 @@
...
+                      String.class,
+                      boolean.class))
+              .bindTo(this);
+
+      List<Procedure.Argument> arguments =
+          List.of(
+              new Procedure.Argument("METALAKE", VARCHAR),
+              new Procedure.Argument("CATALOG", VARCHAR),
+              new Procedure.Argument("PROVIDER", VARCHAR),
+              new Procedure.Argument("PROPERTIES", VARCHAR, false, "{}"),
```
Comment (jerryshao) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:60]: Can it be Map type?

**Comment by diqiu50** [comment]
> So if the configured metalake is "test", and with call procedure you created a new one "test1", can "test1" be used by this connector, or user can still use the metalake "test"?

If 'test' metadata lake is configured and 'test1' metadata lake is not configured, users can still utilize the 'test' metadata lake. Trino will not load the catalog for the 'test1' metadata lake.

The primary purpose of creating a metalake is to allow users to create configured metalakes through Trino. Users can configure multiple Gravitino connectors, each using a different metadata lake.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java (hunk 6 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:60]: I think the map type in Trino a bit challenging to use, the usage is as follows:
```
call gravitino.system.createCatalog('test', 'jdbc-mysql7', 'jdbc-mysql', MAP(ARRAY['jdbc-url', 'jdbc-user', 'jdbc-password','jdbc-dri
ver'], ARRAY['jdbc:mysql://192.168.164.4:3306?useSSL=false', 'mysql', 'mysql', 'com.mysql.cj.jdbc.Driver']));
```
Using a JSON string might align better with our preferences.
```
call gravitino.system.createCatalog('test', 'jdbc-mysql', 'jdbc-mysql', '{"jdbc-url":"jdbc:mysql://192.168.164.4:3306?useSSL=false","jdbc-user":"mysql","jdbc-password":"mysql", "jdbc-driver":"com.mysql.cj.jdbc.Driver"}');
```

**Comment by diqiu50** [comment]
@jerryshao 
I have already removed the stored procedures for creating/dropping the `metalake` and added relevant documentation.

Regarding the attributes section in 'create catalog,' Trino doesn't have a better type of solution to this problem. Using a JSON format or a format similar to 'key1=value1;key2=value2;...' could be relatively more user-friendly.

Later, if we make changes to Trino, we can define specific syntax for the 'create catalog' statement to address the issues caused by using plain strings.

```
CREATE  CATALOG NAME by PROVIDER WITH (
	key1:value1,
	key2:value2,
);
```
Currently, I use JSON, which makes implementation easier.

**Comment by jerryshao** [comment]
@diqiu50 would you please update the PR title and description to reflect the current status of PR?

File: docs/trino-connector/supported-catalog.md [hunk 7]
```diff
@@ -15,7 +15,43 @@ The catalogs currently supported by the Gravitino connector are as follows:
...
+
+Create catalog:
+
+```text
+createCatalog(CATALOG varchar, PROVIDER varchar, PROPERTIES varchar, IGNORE EXIST boolean);
+```
+
+- CATALOG: The catalog name to be created.
+- PROVIDER: The catalog provider, currently only supports `hive`, `lakehouse-iceberg`, `jdbc-mysql`, `jdbc-postgresql`.
+- PROPERTIES: The properties of the catalog, the format is a json string. like `{"key1":"value1","key2":"value2"}`.
```
Comment (jerryshao) [inline docs/trino-connector/supported-catalog.md:30]: Should you also specify which catalog properties are supported or provided by referring to the specific docs?

Comment (jerryshao) [reply docs/trino-connector/supported-catalog.md:30]: Also can you follow this Iceberg doc to make it as a table for properties. https://iceberg.apache.org/docs/latest/spark-procedures/

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java (hunk 6 above)
Comment (jerryshao) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:56]: I found that in https://iceberg.apache.org/docs/latest/spark-procedures/#examples-8, there're some call procedures that have required properties as map type. So I'm inclined to use map type. Though it is not so straight to write, but the definition of map type is much more clear than a vague varchar.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java [hunk 8]
```diff
@@ -20,8 +30,52 @@
...
+          MethodHandles.lookup()
+              .unreflect(
+                  DummyGravitinoConnector.class.getMethod(
+                      "dropCatalog", String.class, boolean.class))
+              .bindTo(this);
+      arguments =
+          List.of(
+              new Procedure.Argument("CATALOG", VARCHAR),
+              new Procedure.Argument("IGNORE NOT EXIST", BOOLEAN, false, true));
+      procedure = new Procedure("system", "dropCatalog", arguments, dropCatalog);
```
Comment (jerryshao) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:72]: Is trino case insensitive, what will be happened if user write as "DROPCATALOG"? I would suggest to use "drop_catalog", "create_catalog".

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConfig.java:53]: What is the usage of this `isStatic()`? You should have a better name or have a better explanation.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java [hunk 9]
```diff
@@ -56,7 +67,7 @@ public class CatalogConnectorManager {
 
   private GravitinoClient gravitinoClient;
   private GravitinoConfig config;
-  private String usedMetalake;
+  private final Set<String> usedMetalakes = new HashSet<>();
```
Comment (jerryshao) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:70]: Is this change still required?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java [hunk 10]
```diff
@@ -191,4 +206,110 @@ public void shutdown() {
...
+      throw new TrinoException(
+          GRAVITINO_CATALOG_ALREADY_EXISTS,
+          "Catalog " + catalog + " already exists in the server.");
+    } catch (Exception e) {
+      throw new TrinoException(
+          GRAVITINO_UNSUPPORTED_OPERATION, "Create catalog failed. " + e.getMessage(), e);
+    }
+  }
+
+  public void createMetalake(String name, boolean ignoreExist) {
```
Comment (jerryshao) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:251]: I think this is not required, right?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java (hunk 9 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:70]: We need to support configuring multiple connectors, each using a different metastore.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java (hunk 10 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:251]: Yes

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java (hunk 8 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:72]: Ok

**Comment by diqiu50** [comment]
@jerryshao 
I have changed the properties to a map type and updated the documents.
Please take a review.

File: docs/trino-connector/supported-catalog.md [hunk 11]
```diff
@@ -15,7 +15,74 @@ The catalogs currently supported by the Gravitino connector are as follows:
...
 ## Create catalog
 
-Trino currently does not support creating Gravitino managed catalogs, if you need to create a catalog, please refer to: [Create a Catalog](../manage-metadata-using-gravitino.md#create-a-catalog).
+Trino itself does not support creating catalogs. 
+Users can create catalogs through the Gravitino connector and then load them into Trino. 
+The Gravitino connector provides the following stored procedures to create and delete catalogs.
+
+Create catalog:
+
+```text
```
Comment (jerryshao) [inline docs/trino-connector/supported-catalog.md:24]: Can you please change this to `sql`?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java [hunk 12]
```diff
@@ -20,8 +33,53 @@
...
+          List.of(
+              new Procedure.Argument("CATALOG", VARCHAR),
+              new Procedure.Argument("PROVIDER", VARCHAR),
+              new Procedure.Argument(
+                  "PROPERTIES", new MapType(VARCHAR, VARCHAR, new TypeOperators())),
+              new Procedure.Argument("IGNORE_EXIST", BOOLEAN, false, true));
+      Procedure procedure = new Procedure("system", "create_catalog", arguments, createCatalog);
+      procedures.add(procedure);
+
+      // call gravitino.system.dropCatalog(metalake, catalog, ignoreNotExist)
```
Comment (jerryshao) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/DummyGravitinoConnector.java:65]: I think you need to update the comment here.

File: docs/trino-connector/supported-catalog.md [hunk 13]
```diff
@@ -52,7 +52,7 @@ So you need to use the following SQL to call them in the `trino-cli`:
 Example:
 You can run the following SQL to create a catalog named `mysql` with `jdbc-mysql` provider.
 
-```text
+```sql
```
Comment (jerryshao) [inline docs/trino-connector/supported-catalog.md:55]: It's not just here, also above, please double check the whole doc.

**Review by jerryshao — APPROVED** [review]
